### DISCUSSION TITLE
Wait for services available

### DISCUSSION THREAD

**Discussion by captnCC** [spatie/laravel-health#70] [Ideas]
# Motivation
Especially when running the Laravel application in multiple docker containers via docker-compose, k8 or similar it's really useful to have a "startup test" that verifies that all parts of the application are up and ruining before disabling the maintenance mode or marking the deployment as success. 
For now we use our own custom package that does similar checks for redis, databases and more but has the ability to differentiate between temporary failures and hard failures. For example an unresolveable database host would be considered temporary till a certain timeout, while an refused connection due to wrong credentials counts as hard and unrecoverable failure.

# Solution
To adapt this package for such a use case a few additions would be necessary:

- Support for multiple sets of checks to run in different phases of the deployment (for example only check database before running migrations)
- A differentiation between hard and temporary failures like describes above (including definitions for how long a temporary failure is allowed to last before being marked as hard)
- A command to run this checks in a loop till all turns green or a hard failure is detected

If this is feature that would be welcomed in the package, I will start working on a PR. 
A separate package could be a solution too, but I think this would require a lot of duplicate code and breaking compatibility with existing checks.

**Comment by freekmurze** [comment]
If these features doesn't involve too much code, docs + tests are provided, and it can be added in a non-breaking way, I'll consider merging it in.

Send a separate PR for the first bullet point, there's already a command to run all the tests ad hoc, I think we'd only need to add an option to only run specific health checks.

Thanks for your work on this!

**Reply by captnCC** [reply]
I created a PR for a simple implementation of the grouping to get started. A pleasure to be involved in your awesome packages!
